lib/rtti.h

```diff
@@ -0,0 +1,181 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef LIB_RTTI_H_
+#define LIB_RTTI_H_
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <type_traits>
+
+namespace RTTI {
+
+using TypeId = std::uint64_t;
+static constexpr TypeId InvalidTypeId = UINT64_C(0);
+
+namespace detail {
+// Apparently string_view does not optimize properly in constexpr context
+// for gcc < 13
+struct TypeNameHolder {
+    template <size_t N>
+    constexpr TypeNameHolder(const char (&str)[N]) : str(&str[0]), length(N - 1) {}  // NOLINT
+
+    const char *str;
+    size_t length;
+};
+
+template <typename T>
+constexpr TypeNameHolder TypeNameProxy() {
```

Comment:
the p4c mostly has function names start with lowercase

Reply:
Also, please add a comment that we require the T's real type to be in the string so that when someone tries to implement this for MSVC or something they are not too surprised.

Reply:
Sure. I believe for MSVC we'll end with `FUNCSIG` (but I do not have MSVC on hand to check it out). With C++20 we can use something like `std::source_location::function_name`